### ISSUE TITLE
Cannot Keyframe "Visible Characters" Attribute In The RichText Node

### ISSUE THREAD

**Issue by davidpgil** [godotengine/godot#12675]
I tried keyframing a type-writer text effect in my game for a cut-scene using the RichText node and it does not seem to work. Is this a known issue? I was trying to keyframe the "Visible Characters" attribute.

**Comment by Zylann** [comment]
In 3.0 I managed to do it by setting the track as "continous". The "discrete" mode only sets the value on each keyframe rather than interpolating.

**Comment by davidpgil** [comment]
I'm not sure what you mean by "discrete" vs "continuous". Would you clarify?

The way I ended up doing it, is using a GDScript that shows a letter over a certain amount of time...

Are you saying there is a way to get the effect I am asking about without keyframing "Visible Characters"?

**Comment by Zylann** [comment]
![image](https://user-images.githubusercontent.com/1311555/32418748-b40df3f2-c26f-11e7-955d-310df19f808a.png)

**Comment by davidpgil** [comment]
I was mistaken in thinking I needed to format the issues as questions. This issue I have solved for myself with a clumsy workaround. I brought up this issue as a result of my post-mortem on a project I already completed.

**Comment by akien-mga** [comment]
As outlined by @Zylann, this can be done by keyframing "Visible Characters" and setting the interpolation mode to "Continuous".

**Comment by davidpgil** [comment]
@akien-mga , I just confirmed that changing the visible characters channel to _continuous_ allows keyframing. I hope we can get a warning that _discrete_ wont keyframe or something as a UX improvement.

**Comment by akien-mga** [comment]
Well *discrete* does keyframe, the issue is not the keyframing. It's just that with *discrete* there is no interpolation, so if you put 0% as first keyframe and 100% as second keyframe, it will show nothing and then jump directly to 100% (or it might even not show anything if you're setting the animation as looping and the second keyframe is on the last frame, which used to be skipped when looping and replaced by frame 0).

**Comment by davidpgil** [comment]
@akien-mga , I think this should be keyframeable. In my case I actually wanted that effect of "jagged" or "jumpy" animation. I was using it for typewriting text in a retro style cut-scene. I still think whether the setting is "discrete" or "continuous" one should still be able to have working keyframes. In the case of type writer text, I was using it to show only certain letters over a set interval.

**Comment by akien-mga** [comment]
Well I don't know what "keyframeable" means to you. You can keyframe anything in Godot.
If you want a jagged effect, you should keyframe several frames (e.g. 0%, 5%, 10%, 15%, 20%) and stuff will appear discretely. It should work just fine.

Then yes, it's not very useful as you'd likely want it to be dependent on the number of letters and not a percentage of visibility, but then Visible Characters is just not what you should use, and you should implement per-letter display by script (typically every x frames/at the end of a timer, append one letter to the RTL).

**Comment by davidpgil** [comment]
I mean that **keyframing _Visible Frames_ produces no effect when the interpolation is set to _discrete_.** I am suggesting that the UX be changed to indicate that to the user. I was only able to get my typewriter text to work using only the UI by setting the interpolation to _continuous_.

Percentage of letters visible versus number of letters doesn't matter much over a reasonably small amount of text, and number of letters over a very large amount of text would be overkill in my opinion. I have already solved this issue for myself using scripting, just as you have suggested.

Again, I am just suggesting a UX improvement at this point. As you probably know, FOSS GNU/Linux software often suffers from several UX issues that make people turn to commercial software. I would love to help with the project and fix stuff like this myself but I haven't managed to create a space to try.

I definitely appreciate the work done so far. Just trying to help! Should I open another issue for this request?

**Comment by akien-mga** [comment]
> I mean that keyframing Visible Frames produces no effect when the interpolation is set to discrete. 

That would be surprising, but if it's true (and I have no reason not to believe you), that would be a bug. I'll give it a try tonight.

**Comment by tom-jk** [comment]
To summarise this thread as I understand it: animating `visible characters` property (I assume "visible _frames_" was a typo) with an `animationplayer` to start at 0 and end at all characters produces no animation between start and end keys with `discrete` update mode. The OP considers this a UX issue and thinks Godot should warn the user about this.

(OP also thought that "keyframing... produces no effect when the interpolate is set to discrete" but as far I can tell this is either not so, no longer so, or part of the misunderstanding about what discrete mode does. Here is a little demo project just to show things working correctly: [12675.zip](https://github.com/godotengine/godot/files/5002048/12675.zip))

My thoughts: I think the real UX issue if any is that animation tracks for `visible characters` default to `discrete` update mode when added; imo it's more probable that a user wants to interpolate smoothly between keys. A user who doesn't understand what the update mode is or that it's there at all can still emulate discrete-style animation in `continuous` mode with multiple keys of constant value, but can't (without effort) produce continuous-style animation in `discrete` mode.

It looks like integer-type properties default to discrete and float-type to continuous. I've had a brief look at potentially relevant issues/PRs but I don't see a specific rationale for this. Integer-types do work fine in continuous mode.

In conclusion, :man_shrugging:

**Comment by KoBeWi** [comment]
I checked the most common discrete usage, Sprite's frames. There is no difference between discrete and continuous. Although with continuous you are able to put 2 keyframes (first frame and last frame) and they will be nicely interpolated.

On the other hand we have `frame_coords`, which surprisingly don't work well with continuous mode, which is default. At least in 3.x, in 4.0 that property is Vector2i. Also in 4.0 when you use continuous mode with `frame_coords`, you get slightly different results than discrete.

Another thing is that discrete mode is the same as continuous with nearest interpolation
![image](https://user-images.githubusercontent.com/2223172/145997069-b1b902ba-9e1d-462e-8228-9fe1e758c81c.png)

The solution would be to re-evaluate what default update modes are used for what type. We could change discrete to continuous for integer types; it's just a matter of what is more expected. Also we might consider removing the discrete mode and use the nearest interpolation instead.
